File: backend/app/schemas/scenario_modeling.py
"""Scenario modeling schemas for API request/response validation."""

from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field, ConfigDict, validator
from datetime import datetime
from enum import Enum

from app.models.scenario_modeling import ScenarioType, OptimizationObjective


# Request Schemas
class ScenarioModelingRequest(BaseModel):
    """Request schema for creating scenario modeling configuration."""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    template_id: str = Field(..., min_length=1)
    scenario_type: ScenarioType
    optimization_objective: OptimizationObjective
    
    # Configuration parameters
    base_configuration: Dict[str, Any] = Field(default_factory=dict)
    variable_parameters: Dict[str, Any] = Field(default_factory=dict)
    constraint_parameters: Dict[str, Any] = Field(default_factory=dict)
    
    # Context information
    business_context: Dict[str, Any] = Field(default_factory=dict)
    target_audience: Dict[str, Any] = Field(default_factory=dict)
    industry_data: Dict[str, Any] = Field(default_factory=dict)


class ScenarioGenerationRequest(BaseModel):
    """Request schema for generating multiple scenarios."""
    configuration_id: str = Field(..., min_length=1)
    number_of_scenarios: int = Field(default=5, ge=1, le=20)
    scenario_diversity: float = Field(default=0.7, ge=0.0, le=1.0)
    include_baseline: bool = Field(default=True)
    optimization_focus: List[str] = Field(default_factory=list)


class ScenarioPredictionRequest(BaseModel):
    """Request schema for generating predictions for scenarios."""
    scenario_ids: List[str] = Field(..., min_items=1)
    prediction_types: List[str] = Field(
        default=["conversion_rate", "revenue", "engagement"],
        min_items=1
    )
    prediction_horizon_days: int = Field(default=30, ge=1, le=365)
    confidence_level: float = Field(default=0.95, ge=0.5, le=0.99)
    include_feature_importance: bool = Field(default=True)


class OptimizationRequest(BaseModel):
    """Request schema for generating optimization recommendations."""
    configuration_id: str = Field(..., min_length=1)
    max_recommendations: int = Field(default=10, ge=1, le=50)
    minimum_impact_threshold: float = Field(default=0.05, ge=0.0, le=1.0)
    effort_constraint: Optional[str] = Field(default=None)  # low, medium, high
    risk_tolerance: float = Field(default=0.5, ge=0.0, le=1.0)


# Response Schemas
class ScenarioModelResponse(BaseModel):
    """Response schema for scenario models."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str
    name: str
    description: Optional[str]
    parameters: Dict[str, Any]
    template_modifications: Dict[str, Any]
    context_variables: Dict[str, Any]
    prediction_results: Optional[Dict[str, Any]]
    confidence_score: Optional[float]
    is_baseline: bool
    is_active: bool
    execution_priority: int
    created_at: datetime
    updated_at: Optional[datetime]


class ScenarioPredictionResponse(BaseModel):
    """Response schema for scenario predictions."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str
    scenario_id: str
    prediction_type: str
    predicted_value: float
    confidence_interval_lower: float
    confidence_interval_upper: float
    confidence_level: float
    baseline_value: Optional[float]
    improvement_percentage: Optional[float]
    statistical_significance: Optional[float]
    prediction_horizon_days: int
    seasonal_adjustment: Optional[float]
    model_version: str
    feature_importance: Optional[Dict[str, Any]]
    created_at: datetime


class OptimizationRecommendationResponse(BaseModel):
    """Response schema for optimization recommendations."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str
    recommendation_type: str
    priority: int
    title: str
    description: str
    rationale: str
    implementation_steps: List[str]
    estimated_effort_hours: Optional[float]
    required_resources: List[str]
    expected_impact: Dict[str, Any]
    risk_assessment: Dict[str, Any]
    success_probability: float
    status: str
    created_at: datetime


class ScenarioModelingConfigurationResponse(BaseModel):
    """Response schema for scenario modeling configuration."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str
    name: str
    description: Optional[str]
    template_id: str
    scenario_type: ScenarioType
    optimization_objective: OptimizationObjective
    base_configuration: Dict[str, Any]
    variable_parameters: Dict[str, Any]
    constraint_parameters: Dict[str, Any]
    business_context: Dict[str, Any]
    target_audience: Dict[str, Any]
    industry_data: Dict[str, Any]
    created_at: datetime
    updated_at: Optional[datetime]
    
    # Computed fields
    scenarios_count: Optional[int] = None
    active_scenarios_count: Optional[int] = None
    latest_prediction_date: Optional[datetime] = None


class ScenarioComparisonResponse(BaseModel):
    """Response schema for scenario comparison analysis."""
    configuration_id: str
    comparison_type: str
    scenarios: List[ScenarioModelResponse]
    predictions: List[ScenarioPredictionResponse]
    
    # Comparison metrics
    best_performing_scenario: Optional[str] = None
    performance_ranking: List[Dict[str, Any]] = Field(default_factory=list)
    statistical_significance_matrix: Dict[str, Dict[str, float]] = Field(default_factory=dict)
    
    # Insights
    key_insights: List[str] = Field(default_factory=list)
    recommended_scenario: Optional[str] = None
    confidence_in_recommendation: Optional[float] = None


class ScenarioSimulationRequest(BaseModel):
    """Request schema for real-time scenario simulation."""
    template_id: str = Field(..., min_length=1)
    modifications: Dict[str, Any] = Field(default_factory=dict)
    
    # Context for simulation
    business_context: Dict[str, Any] = Field(default_factory=dict)
    target_audience: Dict[str, Any] = Field(default_factory=dict)
    market_conditions: Dict[str, Any] = Field(default_factory=dict)
    
    # Simulation parameters
    simulation_duration_days: int = Field(default=30, ge=1, le=365)
    sample_size: int = Field(default=1000, ge=100, le=100000)
    monte_carlo_runs: int = Field(default=1000, ge=100, le=10000)


class ScenarioSimulationResponse(BaseModel):
    """Response schema for scenario simulation results."""
    simulation_id: str
    template_id: str
    modifications: Dict[str, Any]
    
    # Simulation results
    predicted_outcomes: Dict[str, Any]
    confidence_intervals: Dict[str, Dict[str, float]]
    risk_metrics: Dict[str, float]
    
    # Comparison with baseline
    baseline_comparison: Dict[str, float]
    improvement_probability: float
    
    # Detailed analysis
    sensitivity_analysis: Dict[str, float]
    scenario_robustness: float
    implementation_complexity: str  # low, medium, high
    
    # Metadata
    simulation_parameters: Dict[str, Any]
    computation_time_seconds: float
    created_at: datetime


class HistoricalPerformanceRequest(BaseModel):
    """Request schema for historical performance analysis."""
    template_ids: List[str] = Field(..., min_items=1)
    date_range_start: datetime
    date_range_end: datetime
    metrics: List[str] = Field(
        default=["conversion_rate", "engagement", "revenue"],
        min_items=1
    )
    aggregation_level: str = Field(default="daily")  # hourly, daily, weekly, monthly
    include_external_factors: bool = Field(default=True)


class HistoricalPerformanceResponse(BaseModel):
    """Response schema for historical performance data."""
    template_ids: List[str]
    date_range_start: datetime
    date_range_end: datetime
    
    # Performance data
    performance_data: Dict[str, List[Dict[str, Any]]]
    aggregated_metrics: Dict[str, Dict[str, float]]
    
    # Trends and patterns
    trend_analysis: Dict[str, str]  # improving, declining, stable
    seasonal_patterns: Dict[str, List[float]]
    correlation_matrix: Dict[str, Dict[str, float]]
    
    # Insights
    performance_insights: List[str]
    anomaly_detection: List[Dict[str, Any]]
    
    # Metadata
    data_quality_score: float
    created_at: datetime


# Complex Analysis Schemas
class MultiDimensionalAnalysisRequest(BaseModel):
    """Request for multi-dimensional scenario analysis."""
    configuration_id: str = Field(..., min_length=1)
    dimensions: List[str] = Field(..., min_items=2)  # industry, audience, goals, etc.
    interaction_effects: bool = Field(default=True)
    max_combinations: int = Field(default=100, ge=10, le=1000)
    

class MultiDimensionalAnalysisResponse(BaseModel):
    """Response for multi-dimensional analysis results."""
    configuration_id: str
    dimensions_analyzed: List[str]
    
    # Analysis results
    dimension_effects: Dict[str, Dict[str, float]]
    interaction_effects: Dict[str, float]
    optimal_combinations: List[Dict[str, Any]]
    
    # Performance insights
    top_performing_combinations: List[Dict[str, Any]]
    dimension_importance_ranking: List[Dict[str, float]]
    
    # Recommendations
    configuration_recommendations: List[Dict[str, Any]]
    next_experiments: List[Dict[str, Any]]
    
    created_at: datetime

File: web-builder/src/lib/types/scenario-modeling.ts
/**
 * Scenario Modeling Types for Advanced Template Configuration Optimization
 * 
 * These types support comprehensive scenario modeling including:
 * - Multi-dimensional parameter space exploration
 * - Predictive analytics with ML models
 * - Monte Carlo simulations for risk assessment
 * - Statistical significance testing
 * - Real-time optimization recommendations
 */

import { WorkflowCategory } from './smart-templates';

// Core Scenario Modeling Types
export enum ScenarioType {
  TRAFFIC_VOLUME = "traffic_volume",
  AUDIENCE_SEGMENT = "audience_segment",
  BUSINESS_GOAL = "business_goal",
  INDUSTRY_CONTEXT = "industry_context",
  SEASONAL_VARIATION = "seasonal_variation",
  COMPETITIVE_LANDSCAPE = "competitive_landscape",
  MARKETING_BUDGET = "marketing_budget",
  TEMPLATE_VARIATION = "template_variation"
}

export enum OptimizationObjective {
  CONVERSION_RATE = "conversion_rate",
  REVENUE = "revenue",
  ENGAGEMENT = "engagement",
  LEAD_QUALITY = "lead_quality",
  COST_EFFICIENCY = "cost_efficiency",
  USER_RETENTION = "user_retention",
  BRAND_AWARENESS = "brand_awareness"
}

// Configuration and Request Types
export interface ScenarioModelingConfiguration {
  id: string;
  name: string;
  description?: string;
  template_id: string;
  scenario_type: ScenarioType;
  optimization_objective: OptimizationObjective;
  base_configuration: Record<string, any>;
  variable_parameters: Record<string, any>;
  constraint_parameters: Record<string, any>;
  business_context: Record<string, any>;
  target_audience: Record<string, any>;
  industry_data: Record<string, any>;
  created_at: string;
  updated_at?: string;
  
  // Computed fields
  scenarios_count?: number;
  active_scenarios_count?: number;
  latest_prediction_date?: string;
}

export interface ScenarioModelingRequest {
  name: string;
  description?: string;
  template_id: string;
  scenario_type: ScenarioType;
  optimization_objective: OptimizationObjective;
  base_configuration?: Record<string, any>;
  variable_parameters?: Record<string, any>;
  constraint_parameters?: Record<string, any>;
  business_context?: Record<string, any>;
  target_audience?: Record<string, any>;
  industry_data?: Record<string, any>;
}

export interface ScenarioGenerationRequest {
  configuration_id: string;
  number_of_scenarios?: number;
  scenario_diversity?: number;
  include_baseline?: boolean;
  optimization_focus?: string[];
}

// Scenario Model Types
export interface ScenarioModel {
  id: string;
  name: string;
  description?: string;
  configuration_id: string;
  parameters: Record<string, any>;
  template_modifications: Record<string, any>;
  context_variables: Record<string, any>;
  prediction_results?: Record<string, any>;
  confidence_score?: number;
  is_baseline: boolean;
  is_active: boolean;
  execution_priority: number;
  created_at: string;
  updated_at?: string;
}

// Prediction Types
export interface ScenarioPrediction {
  id: string;
  scenario_id: string;
  prediction_type: string;
  predicted_value: number;
  confidence_interval_lower: number;
  confidence_interval_upper: number;
  confidence_level: number;
  baseline_value?: number;
  improvement_percentage?: number;
  statistical_significance?: number;
  prediction_horizon_days: number;
  seasonal_adjustment?: number;
  model_version: string;
  feature_importance?: Record<string, any>;
  created_at: string;
}

export interface PredictionRequest {
  scenario_ids: string[];
  prediction_types?: string[];
  prediction_horizon_days?: number;
  confidence_level?: number;
  include_feature_importance?: boolean;
}

export interface PredictionResult {
  predicted_value: number;
  confidence_interval_lower: number;
  confidence_interval_upper: number;
  model_confidence: number;
  feature_importance: number[];
}

// Optimization Types
export interface OptimizationRecommendation {
  id: string;
  recommendation_type: string;
  priority: number;
  title: string;
  description: string;
  rationale: string;
  implementation_steps: string[];
  estimated_effort_hours?: number;
  required_resources: string[];
  expected_impact: Record<string, any>;
  risk_assessment: Record<string, any>;
  success_probability: number;
  status: string;
  created_at: string;
}

export interface OptimizationRequest {
  configuration_id: string;
  max_recommendations?: number;
  minimum_impact_threshold?: number;
  effort_constraint?: string;
  risk_tolerance?: number;
}

// Simulation Types
export interface ScenarioSimulationRequest {
  template_id: string;
  modifications: Record<string, any>;
  business_context?: Record<string, any>;
  target_audience?: Record<string, any>;
  market_conditions?: Record<string, any>;
  simulation_duration_days?: number;
  sample_size?: number;
  monte_carlo_runs?: number;
}

export interface ScenarioSimulationResult {
  simulation_id: string;
  template_id: string;
  modifications: Record<string, any>;
  predicted_outcomes: Record<string, any>;
  confidence_intervals: Record<string, Record<string, number>>;
  risk_metrics: Record<string, number>;
  baseline_comparison: Record<string, number>;
  improvement_probability: number;
  sensitivity_analysis: Record<string, number>;
  scenario_robustness: number;
  implementation_complexity: 'low' | 'medium' | 'high';
  simulation_parameters: Record<string, any>;
  computation_time_seconds: number;
  created_at: string;
}

// Analysis Types
export interface ScenarioComparison {
  configuration_id: string;
  comparison_type: string;
  scenarios: ScenarioModel[];
  predictions: ScenarioPrediction[];
  best_performing_scenario?: string;
  performance_ranking: Array<{ scenario_id: string; score: number }>;
  statistical_significance_matrix: Record<string, Record<string, number>>;
  key_insights: string[];
  recommended_scenario?: string;
  confidence_in_recommendation?: number;
}

export interface HistoricalPerformanceRequest {
  template_ids: string[];
  date_range_start: string;
  date_range_end: string;
  metrics?: string[];
  aggregation_level?: string;
  include_external_factors?: boolean;
}

export interface HistoricalPerformanceData {
  template_ids: string[];
  date_range_start: string;
  date_range_end: string;
  performance_data: Record<string, Array<Record<string, any>>>;
  aggregated_metrics: Record<string, Record<string, number>>;
  trend_analysis: Record<string, string>;
  seasonal_patterns: Record<string, number[]>;
  correlation_matrix: Record<string, Record<string, number>>;
  performance_insights: string[];
  anomaly_detection: Array<{
    date: string;
    metric: string;
    value: number;
    expected_range: [number, number];
    severity: string;
  }>;
  data_quality_score: number;
  created_at: string;
}

// Multi-dimensional Analysis
export interface MultiDimensionalAnalysisRequest {
  configuration_id: string;
  dimensions: string[];
  interaction_effects?: boolean;
  max_combinations?: number;
}

export interface MultiDimensionalAnalysisResult {
  configuration_id: string;
  dimensions_analyzed: string[];
  dimension_effects: Record<string, Record<string, number>>;
  interaction_effects: Record<string, number>;
  optimal_combinations: Array<{
    combination: Record<string, string>;
    predicted_performance: number;
    confidence: number;
  }>;
  top_performing_combinations: Array<{
    rank: number;
    combination: Record<string, string>;
    performance_score: number;
  }>;
  dimension_importance_ranking: Array<{
    dimension: string;
    importance: number;
  }>;
  configuration_recommendations: Array<{
    recommendation: string;
    expected_lift: number;
    confidence: number;
  }>;
  next_experiments: Array<{
    experiment_name: string;
    parameters: Record<string, any>;
    expected_duration_days: number;
  }>;
  created_at: string;
}

// Component Props Types
export interface ScenarioModelingDashboardProps {
  templateId?: string;
  onConfigurationCreate?: (config: ScenarioModelingConfiguration) => void;
  onScenarioGenerate?: (scenarioIds: string[]) => void;
  onPredictionComplete?: (predictions: ScenarioPrediction[]) => void;
  onRecommendationSelect?: (recommendation: OptimizationRecommendation) => void;
  className?: string;
}

export interface ScenarioConfigurationBuilderProps {
  templateId: string;
  onConfigurationCreate: (config: ScenarioModelingConfiguration) => void;
  onCancel?: () => void;
  initialValues?: Partial<ScenarioModelingRequest>;
}

export interface ScenarioGeneratorProps {
  configurationId: string;
  onScenariosGenerated: (scenarios: ScenarioModel[]) => void;
  onError?: (error: string) => void;
}

export interface PredictionEngineProps {
  scenarios: ScenarioModel[];
  onPredictionsGenerated: (predictions: ScenarioPrediction[]) => void;
  predictionTypes?: string[];
  confidenceLevel?: number;
}

export interface ScenarioComparisonProps {
  configurationId: string;
  scenarios: ScenarioModel[];
  predictions: ScenarioPrediction[];
  onScenarioSelect?: (scenario: ScenarioModel) => void;
  comparisonMetrics?: string[];
}

export interface OptimizationRecommendationsProps {
  configurationId: string;
  recommendations: OptimizationRecommendation[];
  onRecommendationImplement?: (recommendation: OptimizationRecommendation) => void;
  onRecommendationDismiss?: (recommendationId: string) => void;
  filterByPriority?: number;
  filterByStatus?: string;
}

export interface SimulationRunnerProps {
  templateId: string;
  modifications: Record<string, any>;
  onSimulationComplete: (result: ScenarioSimulationResult) => void;
  onSimulationError?: (error: string) => void;
  simulationConfig?: Partial<ScenarioSimulationRequest>;
}

export interface PerformanceAnalyticsProps {
  templateIds: string[];
  dateRange: {
    start: string;
    end: string;
  };
  onAnalysisComplete?: (analysis: HistoricalPerformanceData) => void;
  metrics?: string[];
}

// State Management Types
export interface ScenarioModelingState {
  configurations: ScenarioModelingConfiguration[];
  scenarios: Record<string, ScenarioModel[]>;
  predictions: Record<string, ScenarioPrediction[]>;
  recommendations: Record<string, OptimizationRecommendation[]>;
  simulationResults: Record<string, ScenarioSimulationResult[]>;
  loading: {
    configurations: boolean;
    scenarios: boolean;
    predictions: boolean;
    recommendations: boolean;
    simulation: boolean;
    analysis: boolean;
  };
  errors: {
    configurations?: string;
    scenarios?: string;
    predictions?: string;
    recommendations?: string;
    simulation?: string;
    analysis?: string;
  };
}

export interface ScenarioModelingActions {
  createConfiguration: (request: ScenarioModelingRequest) => Promise<ScenarioModelingConfiguration>;
  generateScenarios: (request: ScenarioGenerationRequest) => Promise<ScenarioModel[]>;
  generatePredictions: (request: PredictionRequest) => Promise<ScenarioPrediction[]>;
  generateRecommendations: (request: OptimizationRequest) => Promise<OptimizationRecommendation[]>;
  runSimulation: (request: ScenarioSimulationRequest) => Promise<ScenarioSimulationResult>;
  compareScenarios: (configurationId: string) => Promise<ScenarioComparison>;
  analyzeHistoricalPerformance: (request: HistoricalPerformanceRequest) => Promise<HistoricalPerformanceData>;
  runMultiDimensionalAnalysis: (request: MultiDimensionalAnalysisRequest) => Promise<MultiDimensionalAnalysisResult>;
  deleteConfiguration: (configurationId: string) => Promise<void>;
}

// Visualization Types
export interface ScenarioVisualizationData {
  type: 'comparison' | 'prediction' | 'sensitivity' | 'risk' | 'performance';
  data: any[];
  labels: string[];
  colors?: string[];
  title: string;
  subtitle?: string;
  xAxisLabel?: string;
  yAxisLabel?: string;
  showLegend?: boolean;
  showTooltip?: boolean;
}

export interface RiskMetrics {
  value_at_risk_5: number;
  conditional_value_at_risk: number;
  risk_adjusted_return: number;
  downside_risk: number;
  volatility: number;
  sharpe_ratio?: number;
}

export interface PerformanceMetrics {
  mean: number;
  median: number;
  std: number;
  min: number;
  max: number;
  percentile_25: number;
  percentile_75: number;
  coefficient_of_variation: number;
}

// API Response Types
export interface ScenarioModelingApiResponse<T = any> {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
  metadata?: Record<string, any>;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// Utility Types
export type ScenarioStatus = 'draft' | 'active' | 'completed' | 'archived';
export type PredictionType = 'conversion_rate' | 'revenue' | 'engagement' | 'lead_quality' | 'cost_efficiency';
export type RecommendationStatus = 'pending' | 'approved' | 'implemented' | 'rejected';
export type SimulationStatus = 'queued' | 'running' | 'completed' | 'failed';
export type AnalysisComplexity = 'basic' | 'intermediate' | 'advanced';
export type ConfidenceLevel = 0.90 | 0.95 | 0.99;
export type TimeHorizon = 7 | 14 | 30 | 60 | 90 | 180 | 365;

// Hook Types
export interface UseScenarioModelingOptions {
  autoRefresh?: boolean;
  refreshInterval?: number;
  enableRealTimeUpdates?: boolean;
  cacheResults?: boolean;
  backgroundSync?: boolean;
}

export interface UseScenarioModelingResult {
  state: ScenarioModelingState;
  actions: ScenarioModelingActions;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
  reset: () => void;
}

// Export utility functions type
export interface ScenarioModelingUtils {
  calculateStatisticalSignificance: (sample1: number[], sample2: number[]) => number;
  generateConfidenceInterval: (values: number[], confidenceLevel: number) => [number, number];
  calculateRiskMetrics: (values: number[]) => RiskMetrics;
  calculatePerformanceMetrics: (values: number[]) => PerformanceMetrics;
  formatPredictionValue: (value: number, type: PredictionType) => string;
  assessImplementationComplexity: (modifications: Record<string, any>) => 'low' | 'medium' | 'high';
  validateScenarioConfiguration: (config: Partial<ScenarioModelingRequest>) => { valid: boolean; errors: string[] };
  generateScenarioName: (type: ScenarioType, parameters: Record<string, any>) => string;
}